xarray/core/netcdftimeindex.py

```diff
@@ -0,0 +1,180 @@
+            end = (date_type(parsed.year, parsed.month + 1, 1) -
+                   timedelta(microseconds=1))
+        return date_type(parsed.year, parsed.month, 1), end
+    if resolution == 'day':
+        start = date_type(parsed.year, parsed.month, parsed.day)
+        return start, start + timedelta(days=1, microseconds=-1)
+    if resolution == 'hour':
+        start = date_type(parsed.year, parsed.month, parsed.day, parsed.hour)
+        return start, start + timedelta(hours=1, microseconds=-1)
+    if resolution == 'minute':
+        start = date_type(parsed.year, parsed.month, parsed.day, parsed.hour,
+                          parsed.minute)
+        return start, start + timedelta(minutes=1, microseconds=-1)
+    if resolution == 'second':
+        start = date_type(parsed.year, parsed.month, parsed.day, parsed.hour,
+                          parsed.minute, parsed.second)
+        return start, start + timedelta(seconds=1, microseconds=-1)
+    else:
+        raise KeyError
+
+
+def get_date_field(datetimes, field):
+    return [getattr(date, field) for date in datetimes]
+
+
+def _field_accessor(name, docstring=None):
+    def f(self):
+        return get_date_field(self._data, name)
+
+    f.__name__ = name
+    f.__doc__ = docstring
+    return property(f)
+
+
+def get_date_type(self):
+    return type(self._data[0])
+
+
+class NetCDFTimeIndex(pd.Index):
+    def __new__(cls, data):
+        result = object.__new__(cls)
+        result._data = np.array(data)
+        return result
+
+    year = _field_accessor('year', 'The year of the datetime')
+    month = _field_accessor('month', 'The month of the datetime')
+    day = _field_accessor('day', 'The days of the datetime')
+    hour = _field_accessor('hour', 'The hours of the datetime')
+    minute = _field_accessor('minute', 'The minutes of the datetime')
+    second = _field_accessor('second', 'The seconds of the datetime')
+    microsecond = _field_accessor('microsecond',
+                                  'The microseconds of the datetime')
+    date_type = property(get_date_type)
+
+    def _partial_date_slice(self, resolution, parsed,
+                            use_lhs=True, use_rhs=True):
+        start, end = _parsed_string_to_bounds(self.date_type, resolution,
+                                              parsed)
+        lhs_mask = (self._data >= start) if use_lhs else True
+        rhs_mask = (self._data <= end) if use_rhs else True
+        return (lhs_mask & rhs_mask).nonzero()[0]
+
+    def _get_string_slice(self, key, use_lhs=True, use_rhs=True):
+        parsed, resolution = _parse_iso8601_with_reso(self.date_type, key)
+        loc = self._partial_date_slice(resolution, parsed, use_lhs, use_rhs)
+        return loc
+
+    def get_loc(self, key, method=None, tolerance=None):
+        if isinstance(key, pd.compat.string_types):
+            result = self._get_string_slice(key)
+            # Prevents problem with __contains__ if key corresponds to only
```

Comment:
I lifted the ```__contains__``` function [from pandas](https://github.com/pandas-dev/pandas/blob/master/pandas/tseries/base.py#L233).  It appears that for a non-monotonic DatetimeIndex, this problem also exists (I would expect line 4 to return ```True```):
```
In [1]: import pandas as pd

In [2]: from datetime import datetime

In [3]: non_monotonic = pd.DatetimeIndex([datetime(2000, 1, 1), datetime(1999, 1
   ...: , 1), datetime(2001, 1, 1)])

In [4]: '2000-01-01' in non_monotonic
Out[4]: False

In [5]: '1999-01-01' in non_monotonic
Out[5]: True

In [6]: '2001-01-01' in non_monotonic
Out[6]: True

In [7]: monotonic = pd.DatetimeIndex([datetime(1999, 1, 1), datetime(2000, 1, 1), 
   ...: datetime(2001, 1, 1)])

In [8]: '1999-01-01' in monotonic
Out[8]: True

In [9]: '2000-01-01' in monotonic
Out[9]: True

In [10]: '2001-01-01' in monotonic
Out[10]: True
```
Is it worth discussing with them to see what their recommended fix is?